### DISCUSSION TITLE
Node.js setInterval

### DISCUSSION THREAD

**Discussion by faveoled** [Winterreisender/webview-nodejs#7] [Q&A] [answered]
I noticed that `setInterval` routine stops working after I call `show()` on a webview. It there a way to go around that?

**Comment by Winterreisender** [comment]
Thanks for your feedback! I'll try to figure out the reason later(Possibly next day). Maybe these issues in webview are related:

https://github.com/webview/webview/issues/14
https://github.com/webview/webview/issues/478

**Reply by faveoled** [reply]
Tried calling `w.show()` from within `w.dispatch` callback but the window stops appearing

**Comment by Winterreisender** [comment]
This issue is reproducible, wherever the `setInterval` is, it won't work.
```js
let w = new Webview(true);
w.title("Hello World");
w.size(800,600);
w.navigate("about:blank");
w.dispatch((w)=>{
    setInterval(() => {
        console.log(1);
        w.title("1");
    },1000);
});
w.show();
```

**Comment by Winterreisender** [comment]
webview_deno has the same issue:

```typescript
// deno run -Ar --unstable test.js

import { Webview } from "https://deno.land/x/webview/mod.ts";

const w = new Webview();
w.navigate(`about:blank`);
setInterval(()=>{
    console.log("test log");
},100);
w.run();
```

**Comment by Winterreisender** [comment]
even the nodejs worker won't work
```js
const { Webview } = require('webview-nodejs');
const { isMainThread, workerData,  Worker } = require('worker_threads');

function workerThread() {
    console.log(workerData);
    setInterval(()=>{
        console.log("Hello");
    },100);
}

function mainThread() {
    let w = new Webview(true);
    new Worker(__filename, { workerData: w });
    w.title("Hello World");
    w.size(800,600);
    w.navigate("about:blank");
    w.show();
}

if (isMainThread) {
    mainThread();
} else {
    workerThread();
}
```

**Reply by JoCat** [reply]
As far as I remember, it worked if I called webview in a child thread and not in the main thread

**Comment by Winterreisender** [comment, marked as answer]
This could be a workaround:

```js
let w = new Webview(true);
w.navigate("about:blank");
w.bind("foo", (w) => {
    // write you setInterval code here
    console.log("Hello");
    
});
w.init("setInterval(foo,1000)");
w.show();
```

it uses the `setInterval` in browser and call the function which is bound to the Node.js. I tested on my machine and it works.

It may be a design of webview that, once you started the webview, you should consider the webview as the main place to write code, node.js and other backend should only deal with code that could not be done in webview

**Reply by faveoled** [reply]
Yeah, that might be the right approach indeed. Don't know if it's enough for server-heavy usecases though

**Reply by Winterreisender** [reply]
> Yeah, that might be the right approach indeed. Don't know if it's enough for server-heavy usecases though

Well, though I'm maintaining a 'webview' project, I can recommend two similar projects for that case:
- Wry (Similar to webview but written in Rust) and Tauri
- QtWebEngine
- Electron

And, if your backend is heavy, you could consider writing a HTTP server and use the B/S structure in webview.

**Reply by faveoled** [reply]
> and use the B/S structure in webview
What are you referring to here?

**Reply by Winterreisender** [reply]
> > and use the B/S structure in webview
> What are you referring to here?


Sorry for that because I'm not a native English speaker. I mean you can write a server seperately and let webview communicate with your server via HTTP. It's just a suggestion, I'm not an expert in writing large-scale softwares.

---

I click the `Edit` instead of `Reply` and edited you message by mistake, sorry for that.

**Reply by faveoled** [reply]
> write a server seperately and let webview communicate with your server via HTTP.

Or maybe another node process. Thanks
